src/Windows/Avalonia.Win32/DxgiSwapchain/ID3D11Device.cs

```diff
@@ -0,0 +1,26 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Avalonia.Win32.DxgiSwapchain
+{
+    internal unsafe struct ID3D11Device
```

Comment:
Please, use our COM interop codegen instead. It provides at least some safety with reference tracking / disposal

Reply:
You can find usage examples here:
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/Win32Com/win32.idl
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/WinRT/winrt.idl

Reply:
> You can find usage examples here: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/Win32Com/win32.idl https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/WinRT/winrt.idl

Working on this now. If I have trouble with this (I noticed that the generator doesn't appreciate `typedef` as a keyword already) can I ask for help from you?

Reply:
Yes, sure.

Reply:
Note that it's not a MIDL flavor of IDL, just something we've initially created to drive our macOS backend interop and later used for WinRT and some Win32 COM objects. So fancy OLE/ActiveX stuff is not supported, only plain IUnknown-based COM.

Reply:
> Note that it's not a MIDL flavor of IDL, just something we've initially created to drive our macOS backend interop and later used for WinRT and some Win32 COM objects. So fancy OLE/ActiveX stuff is not supported, only plain IUnknown-based COM.

Would be nice to have a generator that can directly consume the WinSDK (and WinAppSDK to come) IDL files directly, rather than having to massage all of these files by hand into the MIDL flavor. 

Before I had only included the smallest parts of the interfaces I actually needed, but now, it looks like this will bring in the entirety of DXGI and D3D11 APIs resulting in a lot of internal classes being pulled in.

Reply:
>Before I had only included the smallest parts of the interfaces I actually needed, but now, it looks like this will bring in the entirety of DXGI and D3D11 APIs resulting in a lot of internal classes being pulled in.

For now you can replace unused methods with `void MethodName_DontUse();` for now. I'll implement some proper way for skipping method slots later.

Reply:
> > Before I had only included the smallest parts of the interfaces I actually needed, but now, it looks like this will bring in the entirety of DXGI and D3D11 APIs resulting in a lot of internal classes being pulled in.
> 
> For now you can replace unused methods with `void MethodName_DontUse();` for now. I'll implement some proper way for skipping method slots later.

The other thing that's popped up is that I can't resize buffers until the reference to the backing texture has been Released. 
Normally I would pTex2D->Release(); to release it. 
But there's no Release method on the IUnknown managed interface?

Reply:
You need to use `Dispose` instead

Reply:
> You need to use `Dispose` instead

Okay, that's fine, that works. 

Some of these DXGI methods return zero to say "You can stop here" not because an error has occurred. 

I don't want to be eating COMExceptions during display enumeration, but that's what's going to happen in the wrappers. 

What can I do about that? 

E.g.: EnumAdapters returns non-zero to say "That's the last adapter" 
![image](https://user-images.githubusercontent.com/37165910/204829295-c9ccf78f-8690-4bb8-afe5-dacac74434ba.png)

Reply:
You can replace `HRESULT` with `int`

Reply:
`int EnumAdapters(ushort Adapter, IntPtr* value);` with a wrapper method can work too.

Reply:
> Please, use our COM interop codegen instead. It provides at least some safety with reference tracking / disposal

This is done. There is now a Dxgi.idl generating a Dxgi.Generated.cs to follow the conventions you pointed to earlier.